realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -2474,3 +2486,480 @@ void TemperatureDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWra
+    buffer.data = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(frame.get_data()));
+
+    if (frame.is<rs2::depth_frame>()) {
+        // Note: fix_depth_scale returns it's second argument
+        buffer = fix_depth_scale(buffer, _depth_scaled_image[stream]);
+    }
+
+    sensor_msgs::ImagePtr img;
+    img = cv_bridge::CvImage(std_msgs::Header(), _encoding.at(stream.first), buffer).toImageMsg();
+    img->width = width;
+    img->height = height;
+    img->is_bigendian = false;
+    img->step = width * bpp;
+    img->header.frame_id = _optical_frame_id.at(stream); // Not sure about this, but probably it does not matter for us.
+    img->header.stamp = ros::Time::now(); // This is a simplification, does not handle the case when _sync_frames == false
+
+    ROS_INFO("[NOMAGIC] Generated frame: (w=%u, h=%u, enc=%s, ptr=%p)",
+          img->width, img->height, img->encoding.c_str(), img->data.data());
+
+    return img;
+}
+
+void BaseRealSenseNode::nomagicResetTemporalFilter()
+{
+    ROS_INFO("[NOMAGIC] Resetting temporal filter's state");
+    // Temporal filter resets after changing any of its parameters (even to the same value).
+    for (const NamedFilter& named_filter : _filters) {
+        if (named_filter._name == "temporal") {
+            auto value = named_filter._filter->get_option(RS2_OPTION_FILTER_SMOOTH_ALPHA);
+            named_filter._filter->set_option(RS2_OPTION_FILTER_SMOOTH_ALPHA, value);
+            return;
+        }
+    }
+}
+
+std::set<stream_index_pair> BaseRealSenseNode::nomagicFindMissingStreamsInFrameset(const rs2::frameset& frameset)
+{
+    auto required = nomagic_expected_streams;
+    for (auto frame : frameset) {
+        auto stream_type = frame.get_profile().stream_type();
+        auto stream_index = frame.get_profile().stream_index();
+        required.erase(std::make_pair(stream_type, stream_index));
+    }
+    return required;
+}
+
+void BaseRealSenseNode::nomagicStoreFramesetForLazyProcessing(rs2::frameset frameset)
+{
+    auto missing = nomagicFindMissingStreamsInFrameset(frameset);
+    if (!missing.empty()) {
+        std::stringstream missingStr;
+        for (auto&& stream : missing) {
+            missingStr << rs2_stream_to_string(stream.first) << stream.second << ", ";
+        }
+        // It is natural to see these on startup before receiving a complete frameset.
+        // If seen later, it indicates a bug in the earlier processing stage, most likely in nomagicMuxerCallback(...)
+        ROS_WARN("[NOMAGIC] Received an incomplete frameset, missing: %s", missingStr.str().c_str());
+        return;
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(nomagic_frameset_queue_mutex);
+        if (missing.empty()) {
+            frameset.keep();
+            nomagic_frameset_queue.push_back(frameset);
+        }
+
+    }
+}
+
+boost::circular_buffer<rs2::frameset> BaseRealSenseNode::nomagicGetNonEmptyFramesetQueue()
```

Comment:
Opt: that could be also a method of the FrameAccumulator I mentioned above.

Reply:
@mikolajz I don't see any FrameAccumulator mentions anywhere. I guess you mean wrapping this queue (and appropriate methods) into a class, right?